Raise specific error classes for each verification (type) failure

nocache opened the pull request:
When request verification fails in a running system, it can be very difficult to figure out why.
This PR adds different error subclasses for each type of verification failure.
Systems verifying requests can use these to determine what to do - eg logging the error class will allow quicker diagnosis of errors.

nocache committed:
Raise specific error classes for each verification (type) failure

yoshdog commented on lib/jwt_signed_request/verify.rb:
```diff
@@ -68,12 +67,22 @@ def claims
-        claims['body_sha'] == Digest::SHA256.hexdigest(request_body) &&
-        verified_headers?
+    def verify_request!
+      unless request.request_method.casecmp(claims['method'].to_s) == 0
```
TIL `casecmp`

yoshdog reviewed (approved):
Looks good to me

thanks for the improvements @nocache

brushbox reviewed (approved):
Nice!

lparry commented on lib/jwt_signed_request/errors.rb:
```diff
@@ -0,0 +1,12 @@
+  RequestVerificationFailedError = Class.new(UnauthorizedRequestError)
+  %w[Method Path Header Body Query].each do |type|
+    module_eval("Request#{type}VerificationFailedError = Class.new(RequestVerificationFailedError)")
+  end
```
Somewhat personal preference, but I dont think the 2 line saving is worth the loss in grepability here vs just listing these explicitly

lparry reviewed (approved):
Another nice improvement

thoiberg commented on lib/jwt_signed_request/verify.rb:
```diff
@@ -68,12 +67,22 @@ def claims
-        verified_headers?
+    def verify_request!
+      unless request.request_method.casecmp(claims['method'].to_s) == 0
+        raise RequestMethodVerificationFailedError, "Request failed method verification"
```
Do you think there's any benefit to extending these error messages to show the expected vs actual? Something like `Expected GET but received POST`?

Having the expected and actual values would have been helpful when troubleshooting the author platform issues, as we were passing around the JWT token through a few services so what it was setting the values to weren't immediately apparent.

thoiberg reviewed (approved):
LGTM 👍

nocache committed:
Just list each error individually

I thought it might make it clearer the other way, but this is probably better for greppability

nocache committed:
Output the expected and received values in the exception
